src/utils/message_entity.rs

```diff
@@ -0,0 +1,116 @@
+use std::string::FromUtf16Error;
+
+use teloxide_core::types::MessageEntity;
+
+#[derive(Debug)]
+pub enum InvokeEntityError {
+    TooShortText { offset: usize, length: usize },
+    FromUTF16(FromUtf16Error),
+}
+
+impl From<FromUtf16Error> for InvokeEntityError {
+    fn from(e: FromUtf16Error) -> Self {
+        Self::FromUTF16(e)
+    }
+}
+
+impl std::fmt::Display for InvokeEntityError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InvokeEntityError::TooShortText { offset, length } => write!(
+                f,
+                "Text is too short for entity with offset {} and length {}",
+                offset, length
+            ),
+            InvokeEntityError::FromUTF16(e) => e.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for InvokeEntityError {}
+
+fn invoke_entity_from_utf16(
+    text: &[u16],
+    entity: &MessageEntity,
+) -> Result<String, InvokeEntityError> {
+    let start = entity.offset;
+    let end = entity.offset + entity.length;
+    if text.len() < end {
+        Err(InvokeEntityError::TooShortText { offset: entity.offset, length: entity.length })
+    } else {
+        String::from_utf16(&text[start..end]).map_err(Into::into)
+    }
+}
+pub fn invoke_entity(text: &str, entity: &MessageEntity) -> Result<String, InvokeEntityError> {
+    let text: Vec<_> = text.encode_utf16().collect();
+    invoke_entity_from_utf16(text.as_slice(), entity)
+}
+pub fn invoke_entities(
+    text: &str,
+    entities: &Vec<MessageEntity>,
+) -> Result<Vec<String>, InvokeEntityError> {
```

Comment:
I think we should return not `String`s, but actual utf-8 offsets.
I imagine a structure like that:
```rust
pub struct MessageEntityRef<'a> {
    message: &str,
    range: Range<usize>,
    kind: &'a MessageEntityKind,
}

impl<'a> MessageEntityRef<'a> {
    fn kind(&self) -> &'a MessageEntityKind { ... }
    fn text(&self) -> &'a str { ... }
    fn range(&self) -> Range<usize> { ... }
    fn start(&self) -> usize { ... }
    fn end(&self) -> usize { ... }
    fn len(&self) -> usize { ... }
}
```

Reply:
IMHO, we need a string of entity as final result. And there is no reason to store that thing as range.
Especially, access to byte index of string is unsafe by design.

Reply:
I disagree. If you want to analyze the entities more precisely, you do need the range information. If you want to (for example) resend message with the same formatting (or maybe edit something), then you need to know the ranges, just knowing the text doesn't allow you to reconstruct the original text & formatting.

> access to byte index of string is unsafe by design

What do you mean by "unsafe by design"?

Reply:
> What do you mean by "unsafe by design"?

https://doc.rust-lang.org/book/ch08-02-strings.html#indexing-into-strings

Reply:
I see no mentions of "unsafe" or "soundness" for that matter.

Reply:
> I think we should return not `String`s, but actual utf-8 offsets. I imagine a structure like that:
> 
> ```rust
> pub struct MessageEntityRef<'a> {
>     message: &str,
>     range: Range<usize>,
>     kind: &'a MessageEntityKind,
> }
> 
> impl<'a> MessageEntityRef<'a> {
>     fn kind(&self) -> &'a MessageEntityKind { ... }
>     fn text(&self) -> &'a str { ... }
>     fn range(&self) -> Range<usize> { ... }
>     fn start(&self) -> usize { ... }
>     fn end(&self) -> usize { ... }
>     fn len(&self) -> usize { ... }
> }
> ```

done